core/src/packet_sender_info.rs

```diff
@@ -0,0 +1,56 @@
+/// struct PacketSenderInfo contains sender information for pakcets in baning_stage's
+/// unprocessed_buffer. It provides metrics for stake weight block packing prioritization.
+///
+use std::{collections::HashMap, net::IpAddr};
+
+#[derive(Debug, Default)]
+pub struct SenderDetailInfo {
+    pub stake: u64,
+    pub packet_count: u64,
+}
+
+#[derive(Debug, Default)]
+pub struct PacketSenderInfo {
+    // Vec of IP of packets in banking_stage buffer, one-to-one mapped, even after shuffling.
+    pub packet_senders_ip: Vec<IpAddr>,
+    // HashMap keyed by IP, with sender's detail info as Value
+    pub senders_detail: HashMap<IpAddr, SenderDetailInfo>,
+    // implied info are:
+    // 1. packet_senders_ip.len() is the number of buffered packets subjected to this round of
+    //    processing;
+    // 2. sender_detail.values.map(|info| info.stake).sum() is the total stakes presented in buffer
+}
+
+impl PacketSenderInfo {
+    pub fn report(&self, id: u32) {
+        let tot_buffered_packet_count = self.packet_senders_ip.len();
+        let tot_buffered_stakes: u64 = self.senders_detail.values().map(|info| info.stake).sum();
+
+        for (ip, sender_info) in self.senders_detail.iter() {
+            let stake_pct = match tot_buffered_stakes == 0 {
+                true => 0.0,
+                false => sender_info.stake as f64 / tot_buffered_stakes as f64,
+            };
+            let transaction_pct = match tot_buffered_packet_count == 0 {
+                true => 0.0,
+                false => sender_info.packet_count as f64 / tot_buffered_packet_count as f64,
+            };
+
+            datapoint_info!(
```

Comment:
This seems like it could be way too many datapoints to send at once if there are a large number of senders.

We may have to find a different method of reporting this info/condensing it